ujt/server/server.py

```diff
@@ -121,28 +172,43 @@ def GetSLIs(self, request, context):
             ]
         )
 
-        slis_output = []
-        slis_at_timestamp = []
-        # In this implementation, the timestamps are inclusive since we generate new SLIs
-        # However, this is not inherent to the proto semantics
-        while start_dt <= end_dt:
-            slis_at_timestamp = generate_new_random_slis(
-                self.node_map, requested_node_names, start_dt, requested_sli_types
-            )
-            slis_output += slis_at_timestamp
-            start_dt += dt.timedelta(seconds=self.reporting_interval)
-
-        # Can we improve this logic? Doesn't seem very elegant.
-        this_last_reported_timestamp = slis_at_timestamp[0].timestamp.ToDatetime()
-        if this_last_reported_timestamp > self.last_reported_timestamp:
-            self.last_reported_timestamp = this_last_reported_timestamp
-            # Update the server's internal Node state
-            sli_name_map = {sli.node_name: sli for sli in slis_at_timestamp}
-            for node_name, node in self.node_map:
-                del node.slis[:]
-                node.slis.extend([sli_name_map[node.name]])
-
-        return server_pb2.GetSLIsResponse(slis=slis_output)
+        # if an instant was requested, use random values
+        # otherwise, use linear interpolation to test change over time feature
+        random_mode = start_dt == end_dt
+
+        output_slis = []
+        for node_name in requested_node_names:
+            for sli_type in requested_sli_types:
+                if node_contains_sli_type(self.node_map[node_name], sli_type):
+                    node_slis = generate_interval_slis(
+                        node_name,
+                        start_dt,
+                        end_dt,
+                        sli_type,
+                        self.reporting_interval,
+                        random_mode=random_mode,
+                        start_value=self.start_end_values[self.start_end_idx][0],
+                        end_value=self.start_end_values[self.start_end_idx][1],
+                    )
+
+                    if node_slis == []:
+                        continue
+
+                    output_slis += node_slis
+
+                    # update the server's internal store of the nodes
+                    this_last_reported_timestamp = node_slis[-1].timestamp.ToDatetime()
```

Comment:
Previously, we iterated through the time range, generating a set of SLIs for all the requested nodes and types.

Now, we iterate through the nodes and types, and generate the entire range for a single node/sli type pair.